File: headphone-dashboard-sample/modules/sql-percentile/sql-percentile.test.js
const test = require("node:test");
const assert = require("node:assert/strict");
const moduleApi = require("./sql-percentile");

test("suggestMappings matches numeric subject fields to numeric database columns", () => {
  const subject = {
    user_id: "U001",
    concha_width_mm: "24.5",
    age: "32",
    comment: "ok"
  };
  const mappings = moduleApi.suggestMappings(subject, [
    { name: "concha_width", numeric: true },
    { name: "age", numeric: true },
    { name: "comment", numeric: false }
  ]);
  assert.equal(mappings[0].dashboardField, "concha_width_mm");
  assert.equal(mappings[0].value, 24.5);
  assert.equal(mappings[0].dbColumn, "concha_width");
  assert.equal(mappings[0].matchReason, "contains");
  assert.deepEqual(mappings[0].candidates[0], {
    dbColumn: "concha_width",
    confidence: 0.88,
    matchReason: "contains"
  });
  assert.equal(mappings[1].dbColumn, "age");
  assert.equal(mappings[1].confidence, 1);
});

test("suggestMappings handles camelCase dashboard fields", () => {
  const mappings = moduleApi.suggestMappings(
    { conchaWidthMm: "25.1", earCanalDepth: "11.2" },
    [
      { name: "concha_width", numeric: true },
      { name: "ear_canal_depth_mm", numeric: true }
    ]
  );
  assert.equal(mappings[0].dbColumn, "concha_width");
  assert.equal(mappings[1].dbColumn, "ear_canal_depth_mm");
  assert.ok(mappings[1].confidence >= 0.88);
});

test("buildPercentileRequest keeps only usable numeric mappings", () => {
  const request = moduleApi.buildPercentileRequest({
    databasePath: "ear.sqlite",
    table: "ear_data",
    cohortFilters: [{ column: "ear_side", operator: "equals", value: "left" }],
    mappings: [
      { dashboardField: "age", dbColumn: "age", value: "30" },
      { dashboardField: "comment", dbColumn: "comment", value: "ok" },
      { dashboardField: "missing", dbColumn: "", value: "12" }
    ]
  });
  assert.deepEqual(request, {
    databasePath: "ear.sqlite",
    table: "ear_data",
    cohortFilters: [{ column: "ear_side", operator: "equals", value: "left" }],
    mappings: [{ dashboardField: "age", dbColumn: "age", value: 30 }]
  });
});

test("buildResultCards prepares a stable display model", () => {
  const cards = moduleApi.buildResultCards([
    {
      dashboardField: "concha_width_mm",
      dbColumn: "concha_width",
      value: 24.5,
      percentile: 82.25,
      sampleSize: 100,
      mean: 22,
      sd: 2.5,
      min: 15,
      max: 30,
      rank: 82
    }
  ]);
  assert.equal(cards[0].title, "concha_width_mm");
  assert.equal(cards[0].percentileLabel, "P82.3");
  assert.equal(cards[0].band, "high");
  assert.deepEqual(cards[0].metrics[0], { label: "当前值", value: "24.50" });
});

test("buildAnalysisPackage includes explanations and summary", () => {
  const results = [{
    dashboardField: "concha_width_mm",
    dbColumn: "concha_width",
    value: 24.5,
    percentile: 82.25,
    sampleSize: 100,
    mean: 22,
    sd: 2.5,
    min: 15,
    max: 30,
    rank: 82,
    cohortFiltered: true
  }];
  const pack = moduleApi.buildAnalysisPackage({
    subject: { name: "U001" },
    mappings: [{ dashboardField: "concha_width_mm", dbColumn: "concha_width" }],
    results,
    cohort: { filterCount: 1 }
  });
  assert.equal(pack.summary.mappedFieldCount, 1);
  assert.equal(pack.summary.resultCount, 1);
  assert.equal(pack.summary.cohortFiltered, true);
  assert.match(pack.explanations[0], /concha_width_mm/);
  assert.match(pack.explanations[0], /P82.3/);
});

test("analysisPackageToCsv and serializeAnalysisPackage export stable artifacts", () => {
  const pack = moduleApi.buildAnalysisPackage({
    createdAt: "2026-06-25T00:00:00.000Z",
    subject: { name: "U001" },
    results: [{
      dashboardField: "width, mm",
      dbColumn: "concha_width",
      value: 24.5,
      percentile: 82.25,
      sampleSize: 100,
      mean: 22,
      sd: 2.5,
      rank: 82
    }]
  });
  const csv = moduleApi.analysisPackageToCsv(pack);
  assert.match(csv, /^field,dbColumn,value,percentile,sampleSize,mean,sd,rank,explanation/);
  assert.match(csv, /"width, mm",concha_width,24.50,82.3,100,22.00,2.50,82/);
  const json = moduleApi.serializeAnalysisPackage(pack);
  assert.equal(JSON.parse(json).subject.name, "U001");
});

File: headphone-dashboard-sample/modules/sql-percentile/sql_percentile_test.py
import sqlite3
import tempfile
import unittest
from pathlib import Path

from sql_percentile import (
    build_filter_where,
    compute_percentile_analysis,
    compute_percentiles,
    read_schema,
    summarize_distribution,
)


class SqlPercentileTest(unittest.TestCase):
    def test_schema_and_percentiles(self):
        with tempfile.TemporaryDirectory() as directory:
            database_path = Path(directory) / "ear.sqlite"
            with sqlite3.connect(database_path) as connection:
                connection.execute("CREATE TABLE ear_data (name TEXT, concha_width REAL, age INTEGER, ear_side TEXT)")
                connection.executemany(
                    "INSERT INTO ear_data VALUES (?, ?, ?, ?)",
                    [("A", 20.0, 20, "left"), ("B", 24.0, 30, "left"), ("C", 28.0, 40, "right")],
                )

            schema = read_schema(database_path)
            self.assertEqual(schema[0]["name"], "ear_data")
            self.assertTrue(next(column for column in schema[0]["columns"] if column["name"] == "concha_width")["numeric"])

            results = compute_percentiles(database_path, "ear_data", [
                {"dashboardField": "concha_width_mm", "dbColumn": "concha_width", "value": 24.0},
            ])
            self.assertEqual(results[0]["sampleSize"], 3)
            self.assertAlmostEqual(results[0]["percentile"], 2 / 3 * 100)
            self.assertEqual(results[0]["rank"], 2)
            self.assertEqual(results[0]["belowCount"], 1)
            self.assertEqual(results[0]["equalCount"], 1)
            self.assertAlmostEqual(results[0]["mean"], 24.0)
            self.assertAlmostEqual(results[0]["sd"], 3.265986323710904)

            filtered_results = compute_percentiles(
                database_path,
                "ear_data",
                [{"dashboardField": "concha_width_mm", "dbColumn": "concha_width", "value": 24.0}],
                [{"column": "ear_side", "operator": "equals", "value": "left"}],
            )
            self.assertTrue(filtered_results[0]["cohortFiltered"])
            self.assertEqual(filtered_results[0]["sampleSize"], 2)
            self.assertAlmostEqual(filtered_results[0]["percentile"], 100.0)

            analysis = compute_percentile_analysis(
                database_path,
                "ear_data",
                [
                    {"dashboardField": "concha_width_mm", "dbColumn": "concha_width", "value": 24.0},
                    {"dashboardField": "bad", "dbColumn": "missing_column", "value": 1},
                ],
            )
            self.assertEqual(analysis["resultCount"], 1)
            self.assertEqual(analysis["warningCount"], 1)
            self.assertIn("字段不存在", analysis["warnings"][0]["message"])

    def test_distribution_summary(self):
        summary = summarize_distribution([10, 20, 20, 40], 20)
        self.assertEqual(summary["sampleSize"], 4)
        self.assertEqual(summary["rank"], 3)
        self.assertEqual(summary["belowCount"], 1)
        self.assertEqual(summary["equalCount"], 2)
        self.assertAlmostEqual(summary["percentile"], 75.0)
        self.assertAlmostEqual(summary["mean"], 22.5)
        self.assertIsNotNone(summary["zScore"])

    def test_build_filter_where_rejects_unknown_columns(self):
        with self.assertRaises(ValueError):
            build_filter_where(
                [{"column": "unsafe", "operator": "equals", "value": "x"}],
                {"ear_side"},
            )


if __name__ == "__main__":
    unittest.main()
